### ISSUE TITLE
Proposal: `input.x()` raise `SilentException` instead of `None` when value is not yet populated

### ISSUE THREAD

**Issue by wch** [posit-dev/py-shiny#68]
tl;dr: `req()` is supposed to be for validation and `@event()` is supposed to be for controlling reactivity, but it turns out that in some use cases `@event()` actually does a better job of validation than `req()`, which would drive users (like me) to use `@event()` when it's not quite appropriate. I propose we make `input.x()` throw a `SilentException` when the value hasn't been populated yet.

I've found myself using the following two constructs in my apps, and I think it will be a bit confusing for users to understand which one is appropriate for which uses cases:

A. `@event(input.x)`
B. `req(input.x())`

These have two different purposes:

* `@event()`: Control the triggering of reactivity
* `req()`: Validate input values.

However, it's easy to mix up the use of `@event()` and `req()`, because they overlap quite a bit in their behavior.

Here are some use cases. The first two involve controlling reactivity; the remaining three involve validation.

1. Run the code when the user clicks on a button. This works with `@event()` and `req()`.
    * Note: it works with `@event(input.x)` because we have special-cased the handling of the action button value 0. If we didn't do that, users would have to write `@event(input.x, ignore_init=True)`
2. Run the code when an input changes. This can be handled by `@event()`. `req()` does not work well because falsiness in Python is too broad, encompassing `[]`, `""`, `0`, and so on.
3. Is `input.x()` a value which will not cause an error? Usually, this means that the value is not `None`. `@event()` works here; `req()` is too broad.
4. Is `input.x()` a useful (non-falsy) value? `req()` usually works here, but not always.

In case 3 above, `@event()` is actually more useful for validating inputs than `req()`, but that's not exactly the purpose of `@event()`; using it that way would cause an Effect or output to _not_ depend on other reactive inputs.

Generally speaking, there's three kinds of validation that we're interested in:

* Is the value populated at all?
* Is the value truthy?
* Does the value satisfy some custom condition, like `input.x() > 10`?

Currently, we're expecting `req()` to do the lifting for all of these. It only does the second one natively; the first and third require extra logic, like `if input.x() is None: req(False)`.

******

Proposal:

* If `input.x()` is called before it has been populated, it throws a `SilentException`.
    * Could add an option like `input.x(throw=False)`, which causes it to return `None` instead of raising an exception.
* Add a method `input.x.ready()`, which returns `False` until the value has been populated from the client; after that it returns `True`.

Then we could simplify code like this:

```py
@output()
@render_text()
def txt():
    if input.n() is None:
       req(False)
    return f"n*2 is {input.n() * 2}"
```

Into this:

```py
@output()
@render_text()
def txt():
    return f"n*2 is {input.n() * 2}"
```

If users want a `None` for not-yet-populated values, then they can check `input.x.ready()`.

This doesn't remove the need for the third kind of validation, as in `if input.x() <= 10: req(False)`, but it's not possible to entirely get away from writing custom logic there.

Another advantage of having `input.n()` throw an exception if not yet ready is that the type can be specified more precisely. Presently, you can specify types like this:

```py
class MyInputs(Inputs):
    n: reactive.Value[Union[int, None]]

def server(input: Inputs, output: Outputs, session: Session):
    input = cast(MyInputs, input)
```

The `None` is necessary because it is a possible value returned by `input.n()`. But if it were to throw an exception when not ready, then we can be sure that it's always an int, which allows us to simplify it to:

```py
class MyInputs(Inputs):
    n: reactive.Value[int]
```
